RecoJets/JetProducers/plugins/FixedGridRhoProducerFastjetFromRecHit.cc

```diff
@@ -0,0 +1,178 @@
+  desc.add<bool>("skipECAL", false);
+  //eThresHB/HE are from RecoParticleFlow/PFClusterProducer/python/particleFlowRecHitHBHE_cfi.py
+  desc.add<std::vector<double> >("eThresHB", {0.1, 0.2, 0.3, 0.3});
+  desc.add<std::vector<double> >("eThresHE", {0.1, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2});
+  desc.add<double>("maxRapidity", 2.5);
+  desc.add<double>("gridSpacing", 0.55);
+  descriptions.addWithDefaultLabel(desc);
+}
+
+FixedGridRhoProducerFastjetFromRecHit::~FixedGridRhoProducerFastjetFromRecHit() = default;
+
+void FixedGridRhoProducerFastjetFromRecHit::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  std::vector<fastjet::PseudoJet> inputs;
+  auto const &thresholds = iSetup.getData(ecalPFRecHitThresholdsToken_);
+  auto const &caloGeometry = iSetup.getData(caloGeometryToken_);
+
+  if (!skipHCAL_) {
+    auto const &hbheRecHits = iEvent.get(hbheRecHitsTag_);
+    inputs.reserve(inputs.size() + hbheRecHits.size());
+    for (const auto &hit : hbheRecHits) {
+      if (passedHcalNoiseCut(hit)) {
+        math::XYZTLorentzVector hitp4(0, 0, 0, 0);
+        getHitP4(hit.id(), hit.energy(), hitp4, caloGeometry);
+        inputs.push_back(fastjet::PseudoJet(hitp4.Px(), hitp4.Py(), hitp4.Pz(), hitp4.E()));
+      }
+    }
+  }
+
+  if (!skipECAL_) {
+    auto const &ebRecHits = iEvent.get(ebRecHitsTag_);
+    inputs.reserve(inputs.size() + ebRecHits.size());
+    for (const auto &hit : ebRecHits) {
+      if (passedEcalNoiseCut(hit, thresholds)) {
+        math::XYZTLorentzVector hitp4(0, 0, 0, 0);
+        getHitP4(hit.id(), hit.energy(), hitp4, caloGeometry);
+        inputs.push_back(fastjet::PseudoJet(hitp4.Px(), hitp4.Py(), hitp4.Pz(), hitp4.E()));
+      }
+    }
+
+    auto const &eeRecHits = iEvent.get(eeRecHitsTag_);
+    inputs.reserve(inputs.size() + eeRecHits.size());
+    for (const auto &hit : eeRecHits) {
+      if (passedEcalNoiseCut(hit, thresholds)) {
+        math::XYZTLorentzVector hitp4(0, 0, 0, 0);
+        getHitP4(hit.id(), hit.energy(), hitp4, caloGeometry);
+        inputs.push_back(fastjet::PseudoJet(hitp4.Px(), hitp4.Py(), hitp4.Pz(), hitp4.E()));
+      }
+    }
+  }
+
+  bge_.set_particles(inputs);
+  iEvent.put(std::make_unique<double>(bge_.rho()));
+}
+
+void FixedGridRhoProducerFastjetFromRecHit::getHitP4(const DetId &detId,
+                                                     float hitE,
+                                                     math::XYZTLorentzVector &hitp4,
+                                                     const CaloGeometry &caloGeometry) {
+  const CaloSubdetectorGeometry *subDetGeom = caloGeometry.getSubdetectorGeometry(detId);
+  if (subDetGeom != nullptr) {
+    const auto &gpPos = subDetGeom->getGeometry(detId)->repPos();
+    double thispt = hitE / cosh(gpPos.eta());
+    double thispx = thispt * cos(gpPos.phi());
+    double thispy = thispt * sin(gpPos.phi());
+    double thispz = thispt * sinh(gpPos.eta());
+    hitp4.SetPxPyPzE(thispx, thispy, thispz, hitE);
```

Comment:
should it not be corrected for beam spot (or identified signal vertex) position?

Reply:
For this rho producer, we are just aiming for a proxy of pile-up. As long as we can construct a variable whose value increases with pile-up, we are fine. So I think it's okay to not correct the recHit positions for beamspot. 

I have addressed the other comments.